src/powerbox/tools.py

```diff
@@ -175,9 +226,167 @@ def _get_binweights(coords, weights, bins, average=True, bin_ave=True, log_bins=
+        """
+        arr += 0.0
+    return arr.view(np.dtype((np.void, arr.dtype.itemsize * arr.shape[-1])))
+
+
+def _sample_coords_interpolate(coords, bins, weights, angular_resolution=0.1):
+    # Grid is regular + can be ordered only in Cartesian coords.
+    field_shape = [len(c) for c in coords]
+    if isinstance(weights, np.ndarray):
+        weights = weights.reshape(field_shape)
+    else:
+        weights = np.ones(field_shape) * weights
+    # To extrapolate at the edges if needed.
+    # Evaluate it on points in angular coords that we then convert to Cartesian.
+    # Number of angular bins for each radius absk on which to calculate the interpolated power when doing the averaging
+    # Larger wavemodes / radii will have more samples in theta
+    # "bins" is always 1D
+    # Max is to set a minimum number of bins for the smaller wavemode bins
+    num_angular_bins = np.array(
+        np.max(
+            [
+                np.round(2 * np.pi * bins / angular_resolution),
+                np.ones_like(bins) * 100,
+            ],
+            axis=0,
+        ),
+        dtype=int,
+    )
+    if len(coords) > 1:
+        phi_n = [np.linspace(0.0, np.pi, n) for n in num_angular_bins]
+        phi_N = [np.linspace(0.0, 2 * np.pi, n) for n in num_angular_bins]
+
+        dims2avg = len(coords) - 1
+        # Angular resolution is same for all dims
+        phi_n = np.concatenate(
+            [
+                np.array(
+                    np.meshgrid(
+                        *([phi_N[i]] + [phi_n[i]] * (dims2avg - 1)), sparse=False
+                    )
+                ).reshape((dims2avg, num_angular_bins[i] ** dims2avg))
+                for i in range(len(bins))
+            ],
+            axis=-1,
+        )
+        r_n = np.concatenate(
+            [[r] * (num_angular_bins[i] ** dims2avg) for i, r in enumerate(bins)]
+        )
+        sample_coords = _spherical2cartesian(r_n, phi_n)
+    else:
+        sample_coords = bins.reshape(1, -1)
+        r_n = bins
+    # Removing sample coords that belong to masked regions
+    # First, we pad the weights by setting the weights to 0 around bins
+    # that had weights of 0 from the start.
+    # This is to avoid interpolating close to 0-weighted regions.
+    # binary_weights = weights < 1e-10
+    # rolled = [np.roll(binary_weights, shift=1, axis=i) for i in range(len(coords))]
+    # padded_weights = np.logical_or.reduce(rolled)
+    # weights[padded_weights] = 0
+    # np.digitize can only be applied on a 1D array of bins
+    # for-loop over each dimension
+    indxs = np.array(
+        [np.digitize(sample_coords[i], coords[i]) for i in range(len(coords))]
+    )
+    masked_bins = np.array(np.where(weights < 1e-10)).reshape(3, -1)
+    void_samples = _asvoid(indxs.T)
+    void_masked = _asvoid(masked_bins.T)
+    to_remove = np.in1d(void_samples, void_masked)
+    sample_coords = sample_coords[:, ~to_remove]
+    r_n = r_n[~to_remove]
```

Comment:
This code is super opaque to me, which is fine, but maybe let's move it to its own function to enable more concrete testing? Also, is the `.reshape(3, -1)` assuming ndim=3?

Reply:
Ah this has been fixed already. Just a typo.